main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -366,55 +366,63 @@ object Ast {
   }
 
   /**
-    * Contains the edges held in a single constraint.
+    * Represents a labelled edge in the labelled graph.
+    * The labels represent predicate nodes that must co-occur for the dependency to be relevant.
     */
-  case class MultiEdge(head: (Name.Pred, Type), positives: Vector[TypedPredicate], negatives: Vector[TypedPredicate])
+  case class LabelledEdge(head: PredicateNode, p: Polarity, labels: Vector[PredicateNode], body: PredicateNode, loc: SourceLocation)
 
   /**
-    * Represents a body predicate on a MultiEdge.
+    * Represents a node in the labelled graph.
     */
-  case class TypedPredicate(atom: Name.Pred, tpe: Type, loc: SourceLocation)
+  case class PredicateNode(pred: Name.Pred, tpe: Type)
```

Comment:
I would rename this to Label and include the arity explicitly.

Reply:
(I know the arity can be extracted from the tpe, but I would make it an explicit part)

Reply:
Regarding the naming, is is also used for head and body (since they also need to know their arity) so its not only labels